File: TimeFunctions.java
// Time Complexity :  O(N)
// Space Complexity : O(N/2)
// Did this code successfully run on Leetcode :Yes

import java.util.List;
import java.util.Stack;

public class TimeFunctions {
    public int[] exclusiveTime(int n, List<String> logs) {
      Stack<String[]> stack = new Stack<String[]>();
      int[] result = new int[n];
      int prevTime=0;
      for(int i=0; i<logs.size(); i++){
        String[] words = logs.get(i).split(":", -1);
        if(words[1].equals("start")){

          if(!stack.empty()) result[Integer.parseInt(stack.peek()[0])] += Integer.parseInt(words[2])-prevTime;
          prevTime= Integer.parseInt(words[2]);
          stack.push(words);
        }
        else {

          result[Integer.parseInt(stack.peek()[0])] += Integer.parseInt(words[2])+1-prevTime;
          stack.pop();
          prevTime=Integer.parseInt(words[2])+1;
        }
      }
      return result;
    }
  }

File: Time_Functions.py
# Time complexity : O(n) --> n - length of logs
# Space complexity : O(n) --> n - length of logs
# Leetcode : Solved and submitted

class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
      # declare a res array with all 0's
        res = [0]*n
        if not logs:
            return [0]
          
        # setting th curr, prev and empty stack
        curr = prev = 0
        st = []
        
        # traversing each log
        for log in logs:
          # split the log into each data
            data = log.split(':')
            task = int(data[0])
            curr = int(data[2])
            
            # when we have a start log
            if data[1] == 'start':
              # update the res for task on top of the stack
                if st:
                    res[st[-1]] += curr - prev
                # append the new task and update the prev
                st.append(task)
                prev = curr
                
            else:
              # pop the top of the curr and update it in the res
                res[st.pop()] += curr +1 - prev
                # update the prev to effective curr + 1
                prev = curr + 1
        
        return res
